core/crypto/rng_hw.c

```diff
@@ -3,19 +3,89 @@
 
 #include <compiler.h>
 #include <crypto/crypto.h>
+#include <kernel/mutex.h>
 #include <kernel/panic.h>
+#include <malloc.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 #include <types_ext.h>
 
-/* This is a HW RNG, no need for seeding */
+static uint8_t *rng_fifo;
+static size_t rng_fifo_size;
+static size_t rng_fifo_pos;
+static struct mutex rng_fifo_mutex = MUTEX_INITIALIZER;
+
+TEE_Result __weak hw_get_max_available_entropy(size_t *blen)
+{
+	*blen = sizeof(size_t);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_available_entropy(void *buf)
+{
+	TEE_Result ret;
+
+	ret = hw_get_random_bytes(buf, sizeof(size_t));
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_random_bytes(void *buf, size_t blen)
+{
+	uint8_t *buffer = (uint8_t *)buf;
+	size_t buffer_pos = 0;
+	TEE_Result ret = TEE_SUCCESS;
+
+	mutex_lock(&rng_fifo_mutex);
+
+	while (buffer_pos < blen) {
+		/* Refill our FIFO */
+		if (rng_fifo_pos == 0) {
+			while (true) {
+				ret = hw_get_available_entropy(rng_fifo);
+				if (ret == TEE_SUCCESS)
+					break;
+				else if (ret != TEE_ERROR_BUSY)
+					goto out;
+			}
+		}
+
+		buffer[buffer_pos++] = rng_fifo[rng_fifo_pos++];
+		if (rng_fifo_pos == rng_fifo_size)
+			rng_fifo_pos = 0;
+	}
```

Comment:
The simplest implementation of `hw_get_random_bytes()` over `hw_get_available_entropy()` is rather;
```c
TEE_Result __weak hw_get_random_bytes(void *buf, size_t blen)
{
	TEE_Result res = TEE_ERROR_GENERIC;
	uint8_t *ptr = buf;
	size_t rem = blen;
	size_t s = 0;

	while (rem) {
		s = rem;
		res = hw_get_available_entropy(ptr, &s);
		if (res)
			return res;

		ptr += s;
		rem -= s;
	}

	return TEE_SUCCESS;
}
```

And let `hw_get_available_entropy()` use an internal spinlock to read already comulated data.

Reply:
If we call this version of `hw_get_random_bytes()` asking for 2 bytes for instance, then `hw_get_available_entropy()` would have to either throw out the extra entropy that wasn't requested (HW always reads in bursts of at least 32bits) or buffer the extra for the next call. Former is slower and wastes entropy, the latter all we have done with this "simple" implementation is push the FIFO complexity down into each and every driver instead.

Reply:
I think we can have a generic cache for HW accumulated entropy we don't want to trash.

A RNG driver could expose 2 functions:
- `hw_get_entropy_burst_size()` to get no the full HW RNG FIFO size but rather the size of a random data burst, at HW level.
- `hw_get_available_entropy()` to get one or more burst of entropy from the HW and caching them.

A generic RNG API would be
- An generic internal help function `cache_available_entropy()` to get HW ready entropy and accumulated it in the internal cache.
- `crypto_rng_read_available(void *buf, size_t* sz)` would consume only from the cached random bytes.
It would return an error only if the HW reports it. Would use spinlock protection to access cache buffer.
- `crypto_rng_read()` would loop on calling `crypto_rng_read_available()` until the target buffe ris filled or an erro was returned. No need for spinlock or mutex, since this gets handled from `crypto_rng_read_available()`.

I think the legacy function hw_get_random_byte() could be fully removed. Caller should only go through `crypto_rng_read()`.

Reply:
Sounds like we are getting closer to an alignment here. Couple minor points.

I'm not sure I'm seeing the need for `hw_get_available_entropy()` to return the amount read, it should always be the full burst size if available or TEE_BUSY if not. So `crypto_rng_read()` can loop on that, and SMCCC TRNG can return instantly as needed. Both cache the extra entropy in a common location, but maybe that is what you are suggesting with `cache_available_entropy()`. (I'm ignoring the bcm_hwrng.c driver here since it doesn't use the normal RNG framework for some reason and so needs fixed separately).

Arguably the SMCCC TRNG API is backwards, it should return the amount available (up to a set amount), not request an amount and fail when not instantly available. The issue being for platforms that generate only up to 32bits without blocking, will always fail if we request 64bits (which the Linux kernel does) as we cannot get that much without blocking for more. So even with partial caching, SMCCC will still need to be called twice every time before it returns a success the second time. Not much we can do about that now, just an observation.

> I think the legacy function hw_get_random_byte() should be fully removed. Callers should only go through crypto_rng_read().

Agree, that is what this series already does. If we are aligned on this, then this series is still the base work that needs to be done either way to get these two tasks done. Then the non-blocking API can be added in the next series on top of this (this one is already longer than like so I don't want to combine them if I can help it).

Reply:
> Sounds like we are getting closer to an alignment here. Couple minor points.

Yes. Thanks for the discussion. Rather go in the same direction.

> Arguably the SMCCC TRNG API is backwards, it should return the amount available (up to a set amount), not request an amount and fail when not instantly available.

Sorry, I was not clear. Getting available entropy can return 0 bytes without failing.
This is why it provides a buffer of a given size (passed as `len` input argument value) and on return knows how many bytes where found (passed as `len` output argument value).